### ISSUE TITLE
MQTT Revisited

### ISSUE THREAD

**Issue by HerbOtto** [HomeSpan/HomeSpan#23]
I have integrated MQTT (PubSubClient) w/HomeSpan by delaying it's setup during initial boot.  Things work great for 12 or so hours then HomeSpan becomes unresponsive and does not recover.  However the MQTT comms remain functional.  

Question: **Is there a way to add a hook [or tell me a good place to look in the HS code] in order to have HomeSpan and MQTT coexist peacefully?**

I tried different variations of ESP32 thinking it was a connectivity issue w/the HW my latest being the ESP32E and there appears to be no difference - it's a SW conflict.  Somehow MQTT is interfering w/HomeSpan [even tho it's on port 1883] and HS is not reporting anything in the logs.  MQTT is a major requirement for me as to have the ability to integrate w/Home Assistant along w/other controllers throughout the home [both incoming and outgoing messages].  I'm able to easily control the shades from MQTT and have their status be reflected on HomeKit.  Sweet!

I've also tried running MQTT on different cores or on the same core under another process.  This made no difference.
I'm running 6 large vertical shades and a customized SomfyRemote library using the CC1101 and have 20 allocated connections w/my setup.  The custom variant uses nvs_flash rather than the original EEPROM to avoid storage conflicts.

**Comment by HomeSpan** [comment]
I've not used MQTT before, though I have some familiarity with the concepts.  Can you send me your sketch - I can take a look  (and try to reproduce if possible on my own kit).  The fact that it works for some time but then fails suggests a slow memory leak. 

By coincidence, when adding the new method `homeSpan.setPortNum(uint16_t)` (see #20) I found that I failed to specify the  number of simultaneous WiFi connections used for the HomeSpan WiFiServer to 8 (which is how many WiFi Clients are normally used by HomeSpan).  As a result, WiFiServer used its default of 4, which likely meant some internal juggling as HomeSpan tried to keep 8 connections at the same time.   The code has been updated to make sure the number of connections specified in initializing WiFiServer now match the number of clients expected.

It's unlikely this is the causing your MQTT issue, but on the off-chance that it is related, can you download the **development** branch and re-test your MQTT code?

**Comment by HerbOtto** [comment]
I'll try the development branch and let you know - if I still have issues I'll send you the code [which is quite large].  

I did set
`homeSpan.setMaxConnections(20);		// defaults to 8`
in my code since I have a ton of devices accessing HomeKit [homepods, watches, phones, etc]
Are you saying this was ignored?  That may be an issue but I noticed the failure on multiple devices [which means some are not freeing up?]

**Comment by HomeSpan** [comment]
The `homeSpan.setMaxConnections(20)` would not have been ignored - it properly allocates all the memory needed to track  20 WiFi Clients from within the HomeSpan code.  However, I failed to recognize that at a lower level, the WiFiServer constructor itself also has a parameter to specify the number of simultaneous connections.  There is no parameter for this in the Arduino version of the WiFiServer but it exists (as an optional value) in the ESP32 version.  I'm not sure what the ESP32 WiFi Server library does with this, but since its default was 4, I now set it to match HomeSpan's maxConnections as well.

I have over 1 dozen clients at home and have not had any problems running HomeSpan with a max of only 8 connections, but this only means HomeSpan is properly balancing all the WiFi Clients at the application level.  The fact that everything worked even though WiFiServer was maxed out at 4 simultaneous connections may have only been by luck, and your MQTT code exposed this problem.   As noted, I'd be surprised if this is the issue, but maybe we'll get lucky and it is.
